In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("https://static.bc-edx.com/data/dla-1-2/m21/lms/starter/charity_data.csv")
application_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34299 entries, 0 to 34298
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   EIN                     34299 non-null  int64 
 1   NAME                    34299 non-null  object
 2   APPLICATION_TYPE        34299 non-null  object
 3   AFFILIATION             34299 non-null  object
 4   CLASSIFICATION          34299 non-null  object
 5   USE_CASE                34299 non-null  object
 6   ORGANIZATION            34299 non-null  object
 7   STATUS                  34299 non-null  int64 
 8   INCOME_AMT              34299 non-null  object
 9   SPECIAL_CONSIDERATIONS  34299 non-null  object
 10  ASK_AMT                 34299 non-null  int64 
 11  IS_SUCCESSFUL           34299 non-null  int64 
dtypes: int64(4), object(8)
memory usage: 3.1+ MB


In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(columns=['EIN', 'NAME'])
application_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [3]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = ['T9', 'T13', 'T12', 'T2', 'T25', 'T14', 'T29', 'T15', 'T17']
#cutoff value is <500

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: count, dtype: int64

In [4]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace = ['C7000', 'C1700', 'C4000', 'C5000', 'C1270', 'C2700', 'C2800', 'C7100', 'C1300', 'C1280', 'C1230',
                              'C1400', 'C7200', 'C2300', 'C1240', 'C8000', 'C7120', 'C1500', 'C1800', 'C6000', 'C1250', 'C8200', 
                              'C1238', 'C1278', 'C1235', 'C1237', 'C7210', 'C2400', 'C1720', 'C4100', 'C1257', 'C1600', 'C1260', 
                              'C2710', 'C0', 'C3200', 'C1234', 'C1246', 'C1267', 'C1256', 'C1370', 'C1236', 'C2190', 'C4200'
                              'C5200', 'C1248', 'C2600', 'C1580', 'C6100', 'C1820', 'C1900', 'C2570', 'C1570', 'C3700', 'C2500', 
                              'C1283', 'C2380', 'C1732', 'C1728', 'C2170', 'C4120', 'C8210', 'C4500', 'C1245', 'C2561', 'C2150', 
                              'C4200', 'C5200']

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: count, dtype: int64

In [5]:
# Convert categorical data to numeric with `pd.get_dummies`
application_dummies = pd.get_dummies(application_df[['APPLICATION_TYPE', 'AFFILIATION', 'CLASSIFICATION', 'USE_CASE', 'ORGANIZATION', 'INCOME_AMT', 'SPECIAL_CONSIDERATIONS']], dtype='int')
application_dummies.head()

APPLICATION_TYPE_Other  APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  \
0                       0                     1                     0   
1                       0                     0                     0   
2                       0                     0                     0   
3                       0                     0                     0   
4                       0                     0                     0   

   APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  \
0                    0                    0                    0   
1                    1                    0                    0   
2                    0                    0                    1   
3                    1                    0                    0   
4                    1                    0                    0   

   APPLICATION_TYPE_T6  APPLICATION_TYPE_T7  APPLICATION_TYPE_T8  \
0                    0                    0                    0   
1                    0                    0                    0   
2                    0                    0                    0   
3                    0                    0                    0   
4                    0                    0                    0   

   AFFILIATION_CompanySponsored  ...  INCOME_AMT_1-9999  \
0                             0  ...                  0   
1                             0  ...                  1   
2                             1  ...                  0   
3                             1  ...                  0   
4                             0  ...                  0   

   INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
0                       0                         0                   0   
1                       0                         0                   0   
2                       0                         0                   0   
3                       1                         0                   0   
4                       0                         1                   0   

   INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  INCOME_AMT_50M+  \
0                 0                       0                0   
1                 0                       0                0   
2                 0                       0                0   
3                 0                       0                0   
4                 0                       0                0   

   INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  
0                  0                         1                         0  
1                  0                         1                         0  
2                  0                         1                         0  
3                  0                         1                         0  
4                  0                         1                         0  

[5 rows x 41 columns]

In [6]:
application_dummies['STATUS'] = application_df['STATUS']
application_dummies['ASK_AMT'] = application_df['ASK_AMT']
application_dummies.head()

APPLICATION_TYPE_Other  APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  \
0                       0                     1                     0   
1                       0                     0                     0   
2                       0                     0                     0   
3                       0                     0                     0   
4                       0                     0                     0   

   APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  \
0                    0                    0                    0   
1                    1                    0                    0   
2                    0                    0                    1   
3                    1                    0                    0   
4                    1                    0                    0   

   APPLICATION_TYPE_T6  APPLICATION_TYPE_T7  APPLICATION_TYPE_T8  \
0                    0                    0                    0   
1                    0                    0                    0   
2                    0                    0                    0   
3                    0                    0                    0   
4                    0                    0                    0   

   AFFILIATION_CompanySponsored  ...  INCOME_AMT_100000-499999  \
0                             0  ...                         0   
1                             0  ...                         0   
2                             1  ...                         0   
3                             1  ...                         0   
4                             0  ...                         1   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                   0                 0                       0   
1                   0                 0                       0   
2                   0                 0                       0   
3                   0                 0                       0   
4                   0                 0                       0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0                0                  0                         1   
1                0                  0                         1   
2                0                  0                         1   
3                0                  0                         1   
4                0                  0                         1   

   SPECIAL_CONSIDERATIONS_Y  STATUS  ASK_AMT  
0                         0       1     5000  
1                         0       1   108590  
2                         0       1     5000  
3                         0       1     6692  
4                         0       1   142590  

[5 rows x 43 columns]

In [7]:
# Split our preprocessed data into our features and target arrays
X = application_dummies
y = application_df['IS_SUCCESSFUL']

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y)

In [8]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Optimization Attempt #1
Change activation functions

In [35]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=8, activation="sigmoid", input_dim=43))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=8, activation="sigmoid"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary() 

Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_21 (Dense)            (None, 8)                 352       
                                                                 
 dense_22 (Dense)            (None, 8)                 72        
                                                                 
 dense_23 (Dense)            (None, 1)                 9         
                                                                 
Total params: 433 (1.69 KB)
Trainable params: 433 (1.69 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [36]:
# Compile the Sequential model together and customise metrics
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [37]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 3s 2ms/step - loss: 0.5989 - accuracy: 0.7030
Epoch 2/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5633 - accuracy: 0.7277
Epoch 3/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5577 - accuracy: 0.7288
Epoch 4/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5548 - accuracy: 0.7310
Epoch 5/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5530 - accuracy: 0.7303
Epoch 6/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5519 - accuracy: 0.7309
Epoch 7/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5511 - accuracy: 0.7320
Epoch 8/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5508 - accuracy: 0.7327
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5500 - accuracy: 0.7318
Epoch 10/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5493 - accura

804/804 [==============================] - 2s 2ms/step - loss: 0.5423 - accuracy: 0.7367
Epoch 82/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5419 - accuracy: 0.7371
Epoch 83/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5422 - accuracy: 0.7367
Epoch 84/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5422 - accuracy: 0.7365
Epoch 85/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5422 - accuracy: 0.7358
Epoch 86/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5420 - accuracy: 0.7369
Epoch 87/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5420 - accuracy: 0.7361
Epoch 88/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5422 - accuracy: 0.7353
Epoch 89/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5419 - accuracy: 0.7378
Epoch 90/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5416 - accuracy: 

In [38]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5568 - accuracy: 0.7249 - 493ms/epoch - 2ms/step
Loss: 0.5567543506622314, Accuracy: 0.7248979806900024


Out of Relu, Tanh, Linear and Sigmoid, I was not able to find a combination that provided greater than 73% accuracy. Linear seemed promising in the first few epochs but did no better than Relu. Sigmoid has the best accuracy by a small margin, so we will work with that going forward. 
Tanh - 72.0%
Linear - 72.0%
Sigmoid - 72.5%
Relu - 72.2%

## Optimization Attempt #2
Add more neurons to the layers

In [39]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=43, activation="sigmoid", input_dim=43))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=20, activation="sigmoid"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential_8"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_24 (Dense)            (None, 43)                1892      
                                                                 
 dense_25 (Dense)            (None, 20)                880       
                                                                 
 dense_26 (Dense)            (None, 1)                 21        
                                                                 
Total params: 2793 (10.91 KB)
Trainable params: 2793 (10.91 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [40]:
# Compile the Sequential model together and customise metrics
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [41]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 3s 2ms/step - loss: 0.6083 - accuracy: 0.6950
Epoch 2/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5737 - accuracy: 0.7246
Epoch 3/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5694 - accuracy: 0.7254
Epoch 4/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5647 - accuracy: 0.7274
Epoch 5/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5611 - accuracy: 0.7292
Epoch 6/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5579 - accuracy: 0.7301
Epoch 7/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5562 - accuracy: 0.7311
Epoch 8/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5546 - accuracy: 0.7304
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5532 - accuracy: 0.7319
Epoch 10/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5525 - accura

804/804 [==============================] - 2s 2ms/step - loss: 0.5371 - accuracy: 0.7390
Epoch 82/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5367 - accuracy: 0.7396
Epoch 83/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5366 - accuracy: 0.7402
Epoch 84/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5364 - accuracy: 0.7405
Epoch 85/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5365 - accuracy: 0.7397
Epoch 86/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5364 - accuracy: 0.7405
Epoch 87/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5363 - accuracy: 0.7401
Epoch 88/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5361 - accuracy: 0.7389
Epoch 89/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5361 - accuracy: 0.7397
Epoch 90/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5365 - accuracy: 

In [42]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5575 - accuracy: 0.7244 - 528ms/epoch - 2ms/step
Loss: 0.5575137734413147, Accuracy: 0.7244315147399902


Adding more neurons to the layers has not increased accuracy.

## Optimization Attempt #3
Add more layers

In [43]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=8, activation="sigmoid", input_dim=43))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=8, activation="sigmoid"))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=8, activation="sigmoid"))

# Fourth hidden layer
nn.add(tf.keras.layers.Dense(units=8, activation="sigmoid"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential_9"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_27 (Dense)            (None, 8)                 352       
                                                                 
 dense_28 (Dense)            (None, 8)                 72        
                                                                 
 dense_29 (Dense)            (None, 8)                 72        
                                                                 
 dense_30 (Dense)            (None, 8)                 72        
                                                                 
 dense_31 (Dense)            (None, 1)                 9         
                                                                 
Total params: 577 (2.25 KB)
Trainable params: 577 (2.25 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [44]:
# Compile the Sequential model together and customise metrics
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [45]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 3s 3ms/step - loss: 0.6781 - accuracy: 0.5834
Epoch 2/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5810 - accuracy: 0.7307
Epoch 3/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5751 - accuracy: 0.7312
Epoch 4/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5734 - accuracy: 0.7320
Epoch 5/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5727 - accuracy: 0.7321
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5724 - accuracy: 0.7325
Epoch 7/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5721 - accuracy: 0.7315
Epoch 8/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5718 - accuracy: 0.7312
Epoch 9/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5714 - accuracy: 0.7325
Epoch 10/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5709 - accura

804/804 [==============================] - 2s 3ms/step - loss: 0.5484 - accuracy: 0.7355
Epoch 82/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5486 - accuracy: 0.7339
Epoch 83/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5481 - accuracy: 0.7350
Epoch 84/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5483 - accuracy: 0.7349
Epoch 85/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5484 - accuracy: 0.7344
Epoch 86/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5480 - accuracy: 0.7354
Epoch 87/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5483 - accuracy: 0.7345
Epoch 88/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5480 - accuracy: 0.7348
Epoch 89/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5483 - accuracy: 0.7349
Epoch 90/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5482 - accuracy: 

In [46]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5595 - accuracy: 0.7257 - 516ms/epoch - 2ms/step
Loss: 0.5594671368598938, Accuracy: 0.7257142663002014


Adding extra layers has not changed our accuracy level at all.

In [47]:
# Export our model to HDF5 file
nn.save('AlphabetSoupCharity_Optimisation.h5')